File: common/src/main/kotlin/net/impleri/playerskills/data/SkillsDataLoader.kt
package net.impleri.playerskills.data

import com.google.gson.GsonBuilder
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import net.impleri.playerskills.PlayerSkills
import net.impleri.playerskills.api.Skill
import net.impleri.playerskills.api.TeamMode
import net.impleri.playerskills.skills.basic.BasicSkill
import net.impleri.playerskills.skills.basic.BasicSkillType
import net.impleri.playerskills.skills.numeric.NumericSkill
import net.impleri.playerskills.skills.numeric.NumericSkillType
import net.impleri.playerskills.skills.specialized.SpecializedSkill
import net.impleri.playerskills.skills.specialized.SpecializedSkillType
import net.impleri.playerskills.skills.tiered.TieredSkill
import net.impleri.playerskills.skills.tiered.TieredSkillType
import net.impleri.playerskills.utils.SkillResourceLocation
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.packs.resources.ResourceManager
import net.minecraft.server.packs.resources.SimpleJsonResourceReloadListener
import net.minecraft.util.profiling.ProfilerFiller

class SkillsDataLoader : SimpleJsonResourceReloadListener(Gson, "skills") {
  override fun apply(
    datamap: Map<ResourceLocation, JsonElement>,
    resourceManager: ResourceManager,
    profilerFiller: ProfilerFiller,
  ) {
    datamap.forEach { (name, json) ->
      try {
        parseSkill(name, json)?.let {
          PlayerSkills.LOGGER.info("Registering skill ${it.name}")
          Skill.modify(it)
        }
      } catch (e: Throwable) {
        PlayerSkills.LOGGER.warn("Could not parse skill details for $name")
        e.message?.let { PlayerSkills.LOGGER.error("$it.") }
        e.printStackTrace()
      }
    }
  }

  private fun parseSkill(name: ResourceLocation, json: JsonElement): Skill<*>? {
    val raw = json.asJsonObject

    // Required fields will throw an NPE if they cannot be parsed
    val type = getValue(raw, "type")?.asString?.let { SkillResourceLocation.of(it) }
      ?: throw NullPointerException("Could not detect skill type")

    // Optional fields will be nullish if not found in JSON
    val description = parseValue(raw, "description", { it.asString })
    val changesAllowed = parseValue(raw, "changesAllowed", { it.asInt }) ?: Skill.UNLIMITED_CHANGES
    val (notify, notifyString) = parseNotify(raw)

    return when (type) {
      BasicSkillType.NAME -> {
        val initialValue = parseValue(raw, "initialValue", { it.asBoolean })
        val options = parseOptions(raw) { it.asBoolean }
        val teamMode = parseTeamMode(raw)

        BasicSkill(name, initialValue, description, options, changesAllowed, teamMode, notify, notifyString)
      }

      NumericSkillType.NAME -> {
        val initialValue = parseValue(raw, "initialValue", { it.asDouble })
        val options = parseOptions(raw) { it.asDouble }
        val teamMode = parseTeamMode(raw)

        NumericSkill(name, initialValue, description, options, changesAllowed, teamMode, notify, notifyString)
      }

      TieredSkillType.NAME -> {
        val initialValue = parseValue(raw, "initialValue", { it.asString })
        val options = parseOptions(raw) { it.asString }
        val teamMode = parseTeamMode(raw, TeamMode.pyramid())

        TieredSkill(name, initialValue, description, options, changesAllowed, teamMode, notify, notifyString)
      }

      SpecializedSkillType.NAME -> {
        val initialValue = parseValue(raw, "initialValue", { it.asString })
        val options = parseOptions(raw) { it.asString }
        val teamMode = parseTeamMode(raw, TeamMode.splitEvenly())

        SpecializedSkill(
          name,
          initialValue,
          description,
          options,
          changesAllowed,
          teamMode,
          notify,
          notifyString,
        )
      }

      else -> {
        PlayerSkills.LOGGER.warn("Unknown skill type $type for $name")
        null
      }
    }
  }

  private fun getValue(
    raw: JsonObject,
    key: String,
  ): JsonElement? {
    return try {
      raw.get(key)
    } catch (e: NullPointerException) {
      PlayerSkills.LOGGER.info("Could not get value for $key")
      e.message?.let { PlayerSkills.LOGGER.error(it) }
      e.printStackTrace()
      null
    }
  }

  private fun <T> parseValue(
    raw: JsonObject,
    key: String,
    parser: ((element: JsonElement) -> T),
    defaultValue: T? = null,
  ): T? {
    val rawValue = getValue(raw, key) ?: return null

    return if (rawValue.isJsonNull) defaultValue else parser(rawValue)
  }

  private fun <T> parseOptions(
    raw: JsonObject,
    parser: (element: JsonElement) -> T,
  ): List<T>? {
    return parseValue(
      raw,
      "options",
      {
        it.asJsonArray.map { rawValue -> parser(rawValue) }.toList()
      },
    )
  }

  private fun createTeamMode(mode: String, rate: Double? = null): TeamMode {
    return when (mode) {
      "shared" -> TeamMode.shared()
      "splitEvenly" -> TeamMode.splitEvenly()
      "pyramid" -> TeamMode.pyramid()

      "limited" -> rate?.let { TeamMode.limited(it) }
        ?: throw NullPointerException("A rate is required for limited Team Mode")

      "proportional" -> rate?.let { TeamMode.proportional(it) }
        ?: throw NullPointerException("A rate is required for proportional Team Mode")

      else -> TeamMode.off()
    }
  }

  private fun restrictTeamMode(allowed: TeamMode? = null): (TeamMode) -> TeamMode {
    return {
      when {
        it.isPyramid && allowed?.isPyramid != true -> TeamMode.off()
        it.isSplitEvenly && allowed?.isSplitEvenly != true -> TeamMode.off()
        else -> it
      }
    }
  }

  private fun parseTeamMode(
    raw: JsonObject,
    allowed: TeamMode? = null,
  ): TeamMode {
    val rawMode = getValue(raw, "teamMode") ?: return TeamMode.off()

    val restrict = restrictTeamMode(allowed)

    if (rawMode.isJsonNull) {
      return TeamMode.off()
    }

    if (rawMode.isJsonObject) {
      val mode = parseValue(rawMode.asJsonObject, "mode", { it.asString }) ?: ""
      val rate = parseValue(rawMode.asJsonObject, "rate", { it.asDouble })

      return restrict(createTeamMode(mode, rate))
    }

    return restrict(createTeamMode(rawMode.asString, null))
  }

  private fun parseNotify(
    raw: JsonObject,
  ): Pair<Boolean, String?> {
    val notifyString = parseValue(
      raw,
      "notify",
      {
        try {
          it.asString
        } catch (_: Throwable) {
          null
        }
      },
    )

    if (!notifyString.isNullOrEmpty()) {
      return true to notifyString
    }

    val notify = parseValue(
      raw,
      "notify",
      { it.asBoolean },
    ) ?: false

    return notify to null
  }

  companion object {
    private val Gson = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
  }
}

File: common/src/main/scala/net/impleri/playerskills/data/SkillsDataLoader.scala
package net.impleri.playerskills.data

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import net.impleri.playerskills.api.skills.Skill
import net.impleri.playerskills.api.skills.SkillOps
import net.impleri.playerskills.api.skills.TeamMode
import net.impleri.playerskills.data.utils.JsonDataParser
import net.impleri.playerskills.skills.basic.BasicSkill
import net.impleri.playerskills.skills.basic.BasicSkillType
import net.impleri.playerskills.skills.numeric.NumericSkill
import net.impleri.playerskills.skills.numeric.NumericSkillType
import net.impleri.playerskills.skills.specialized.SpecializedSkill
import net.impleri.playerskills.skills.specialized.SpecializedSkillType
import net.impleri.playerskills.skills.tiered.TieredSkill
import net.impleri.playerskills.skills.tiered.TieredSkillType
import net.impleri.playerskills.utils.PlayerSkillsLogger
import net.impleri.playerskills.utils.SkillResourceLocation
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.packs.resources.ResourceManager
import net.minecraft.server.packs.resources.SimpleJsonResourceReloadListener
import net.minecraft.util.profiling.ProfilerFiller

import scala.jdk.javaapi.CollectionConverters
import scala.util.chaining.scalaUtilChainingOps

case class LimitRequiredForTeamMode() extends Exception

case class ProportionRequiredForTeamMode() extends Exception

case class SkillsDataLoader(
  protected val skillOps: SkillOps = Skill(),
  override val logger: PlayerSkillsLogger = PlayerSkillsLogger.SKILLS,
)
  extends SimpleJsonResourceReloadListener(SkillsDataLoader.GsonService, "skills") with JsonDataParser {
  override def apply(
    data: java.util.Map[ResourceLocation, JsonElement],
    resourceManager: ResourceManager,
    profilerFiller: ProfilerFiller,
  ): Unit = {
    CollectionConverters.asScala(data)
      .flatMap(t => parseSkill(t._1, t._2))
      .foreach(skillOps.upsert(_))
  }

  private def parseSkill(name: ResourceLocation, json: JsonElement): Option[Skill[_]] = {
    val raw = json.getAsJsonObject
    val description = parseString(raw, "description")
    val changesAllowed = parseInt(raw, "changesAllowed").getOrElse(Skill.UNLIMITED_CHANGES)
    val (notify, notifyString) = parseNotify(raw)
    val skillType = parseSkillType(raw)

    skillType.flatMap(SkillResourceLocation.apply)
      .flatMap {
        case BasicSkillType.NAME => createBasicSkill(raw, name, description, changesAllowed, notify, notifyString)
        case NumericSkillType.NAME => createNumericSkill(raw, name, description, changesAllowed, notify, notifyString)
        case TieredSkillType.NAME => createTieredSkill(raw, name, description, changesAllowed, notify, notifyString)
        case SpecializedSkillType.NAME => createSpecializedSkill(raw,
          name,
          description,
          changesAllowed,
          notify,
          notifyString,
        )
        case _ =>
        logger.warn(s"Unknown skill $name with type $skillType")
        None
      }
  }

  private def createBasicSkill(
    raw: JsonObject,
    name: ResourceLocation,
    description: Option[String],
    changesAllowed: Int,
    announceChange: Boolean,
    notifyString: Option[String],
  ): Option[BasicSkill] = {
    Option(
      BasicSkill(
        value = parseBoolean(raw, "initialValue"),
        options = parseOptions(raw, castAsBoolean),
        teamMode = parseTeamMode(raw),
        name = name,
        description = description,
        changesAllowed = changesAllowed,
        announceChange = announceChange,
        notifyKey = notifyString,
      ),
    )
  }

  private def createNumericSkill(
    raw: JsonObject,
    name: ResourceLocation,
    description: Option[String],
    changesAllowed: Int,
    announceChange: Boolean,
    notifyString: Option[String],
  ): Option[NumericSkill] = {
    Option(
      NumericSkill(
        value = parseDouble(raw, "initialValue"),
        options = parseOptions(raw, castAsDouble),
        teamMode = parseTeamMode(raw),
        name = name,
        description = description,
        step = parseDouble(raw, "step").getOrElse(NumericSkill.DefaultStep),
        changesAllowed = changesAllowed,
        announceChange = announceChange,
        notifyKey = notifyString,
      ),
    )
  }

  private def createTieredSkill(
    raw: JsonObject,
    name: ResourceLocation,
    description: Option[String],
    changesAllowed: Int,
    announceChange: Boolean,
    notifyString: Option[String],
  ): Option[TieredSkill] = {
    Option(
      TieredSkill(
        value = parseString(raw, "initialValue"),
        options = parseOptions(raw, castAsString),
        teamMode = parseTeamMode(raw, Option(TeamMode.Pyramid())),
        name = name,
        description = description,
        changesAllowed = changesAllowed,
        announceChange = announceChange,
        notifyKey = notifyString,
      ),
    )
  }

  private def createSpecializedSkill(
    raw: JsonObject,
    name: ResourceLocation,
    description: Option[String],
    changesAllowed: Int,
    announceChange: Boolean,
    notifyString: Option[String],
  ): Option[SpecializedSkill] = {
    Option(
      SpecializedSkill(
        value = parseString(raw, "initialValue"),
        options = parseOptions(raw, castAsString),
        teamMode = parseTeamMode(raw, Option(TeamMode.SplitEvenly())),
        name = name,
        description = description,
        changesAllowed = changesAllowed,
        announceChange = announceChange,
        notifyKey = notifyString,
      ),
    )
  }

  private def createTeamMode(mode: String, rate: Option[Double]): Either[Exception, TeamMode] = {
    mode match {
      case "shared" => Right(TeamMode.Shared())
      case "splitEvenly" => Right(TeamMode.SplitEvenly())
      case "pyramid" => Right(TeamMode.Pyramid())
      case "limited" => {
        rate.map(_.floor.toInt)
          .map(TeamMode.Limited)
          .toRight(LimitRequiredForTeamMode())
      }
      case "proportional" => rate.map(TeamMode.Proportional).toRight(ProportionRequiredForTeamMode())
      case _ => Right(TeamMode.Off())
    }
  }

  private def restrictTeamMode(allowed: Option[TeamMode])(mode: TeamMode): TeamMode = {
    mode match {
      case _: TeamMode.Pyramid if !allowed.contains(TeamMode.Pyramid()) => TeamMode.Off()
      case _: TeamMode.SplitEvenly if !allowed.contains(TeamMode.SplitEvenly()) => TeamMode.Off()
      case _ => mode
    }
  }

  private def parseSkillType(raw: JsonObject): Option[String] = {
    parseString(raw, "type").tap {
      case None => logger.warn("Tried to parse undefined skill type")
      case _ => ()
    }
  }

  private def parseTeamMode(
    raw: JsonObject,
    allowed: Option[TeamMode] = None,
  ): TeamMode = {
    parseValue(
      raw,
      "teamMode",
      {
        case m if m.isJsonObject =>
        val mode = parseString(m.getAsJsonObject, "mode")
        val rate = parseDouble(m.getAsJsonObject, "rate")
        mode.flatMap(createTeamMode(_, rate).toOption)
        case s if isPrimitiveType(s, _.isString) => createTeamMode(s.getAsString, None).toOption
        case _ => None
      },
    )
      .map(restrictTeamMode(allowed))
      .getOrElse(TeamMode.Off())
  }

  private def parseNotify(
    raw: JsonObject,
  ): (Boolean, Option[String]) = {
    parseBoolean(raw, "notify") match {
      case Some(v) => (v, None)
      case _ => parseNotifyString(raw)
    }
  }

  private def parseNotifyString(
    raw: JsonObject,
  ): (Boolean, Option[String]) = {
    parseString(raw, "notify") match {
      case Some(v) => (true, Option(v))
      case _ => (false, None)
    }
  }
}

object SkillsDataLoader {
  private[data] val GsonService: Gson = new GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
}
